consensus/hotstuff/signature/packager.go

```diff
@@ -1,26 +1,253 @@
+	}
+
+	// lookup is a map from node identifier to node identity
+	// it is used to check the given signers are all valid signers at the given block
+	lookup := consensus.Lookup()
+
+	for _, stakingSigner := range sig.StakingSigners {
+		_, ok := lookup[stakingSigner]
+		if ok {
+			signerIDs = append(signerIDs, stakingSigner)
+			sigTypes = append(sigTypes, hotstuff.SigTypeStaking)
+		} else {
+			return nil, nil, fmt.Errorf("staking signer ID (%v) not found in the committees at block: %v", stakingSigner, blockID)
+		}
+	}
+
+	for _, beaconSigner := range sig.RandomBeaconSigners {
+		_, ok := lookup[beaconSigner]
+		if ok {
+			signerIDs = append(signerIDs, beaconSigner)
+			sigTypes = append(sigTypes, hotstuff.SigTypeRandomBeacon)
+		} else {
+			return nil, nil, fmt.Errorf("random beacon signer ID (%v) not found in the committees at block: %v", beaconSigner, blockID)
+		}
+	}
+
+	// seralize the sig type for compaction
+	seralized, err := seralizeToBytes(sigTypes)
+	if err != nil {
+		return nil, nil, fmt.Errorf("could not seralize sig types to bytes at block: %v, %w", blockID, err)
+	}
+
+	data := signatureData{
+		SigType:                   seralized,
+		AggregatedStakingSig:      sig.AggregatedStakingSig,
+		AggregatedRandomBeaconSig: sig.AggregatedRandomBeaconSig,
+		RandomBeacon:              sig.ReconstructedRandomBeaconSig,
+	}
+
+	// encode the structured data into raw bytes
+	encoded, err := p.encoder.Encode(data)
+	if err != nil {
+		return nil, nil, fmt.Errorf("could not encode data %v, %w", data, err)
+	}
+
+	return signerIDs, encoded, nil
 }
 
+// Unpack de-serializes the provided signature data.
+// blockID is the block that the aggregated sig is signed for
+// sig is the aggregated signature data
+// It returns:
+//  - (sigData, nil) if successfully unpacked the signature data
+//  - (nil, signature.ErrInvalidFormat) if failed to unpack the signature data
 func (p *ConsensusSigPackerImpl) Unpack(blockID flow.Identifier, signerIDs []flow.Identifier, sigData []byte) (*hotstuff.BlockSignatureData, error) {
-	panic("to be implemented")
+	// decode into typed data
+	var data signatureData
+	err := p.encoder.Decode(sigData, &data)
+	if err != nil {
+		return nil, fmt.Errorf("could not decode sig data %s: %w", err, signature.ErrInvalidFormat)
+	}
+
+	// deseralize the compact sig types
+	sigTypes, err := deseralizeFromBytes(data.SigType, len(signerIDs))
+	if err != nil {
+		return nil, fmt.Errorf("failed to deseralize sig types from bytes: %w", err)
+	}
+
+	// read all the possible signer IDs at the given block
+	consensus, err := p.committees.Identities(blockID, filter.HasRole(flow.RoleConsensus))
+	if err != nil {
+		return nil, fmt.Errorf("could not find consensus committees by block id(%v): %w", blockID, err)
+	}
+
+	// lookup is a map from node identifier to node identity
+	// it is used to check the given signerIDs are all valid signers at the given block
+	lookup := consensus.Lookup()
```

Comment:
this check is added even though it might not be necessary right now, but eventually will be needed once we work on the optimization to use signer indices to replace the signer IDs, as we will need to query the signer index by signer ID